Branch#22

Elpicsida opened the pull request:
The following changes have been introduced in this PR :
- Spell System created - logic separation of  different types of spells , separation between wizardController and casted spells. Spells are now registered as enum and are added in Wizard Template
- Simple model of dealing damage has been created. Spells now contain explosion object which is responsible for dealing damage based on Radius, DamageValue and Position. 
- Movement is now reduced when a wizzard moves left/right.

Elpicsida committed:
Branch #22
- Spell Mechanism added
- Damage dealing based on the distance from explosion
- Condition decreases when the wizzard moves

Elpicsida committed:
Merge branch 'Branch#22' of https://github.com/Elpicsida/Wizards into Branch#22

# Conflicts:
#	Assets/Resources.meta
#	Assets/Scene/TemplateScene.unity
#	Assets/Scripts/Miscellaneous/TurnManager.cs
#	Assets/Scripts/Wizzard/Wizard.cs

Elpicsida committed:
Branch  #22
Merging

Elpicsida committed:
Branch # 22

czkafa commented on Assets/Scripts/Miscellaneous/Explosion.cs:
```diff
@@ -0,0 +1,27 @@
+
+public class Explosion {
+
+    private GameObject[] Wizzards;
```
Wizardow nie bedzie tuzin dlatego nie potrzebny jest array a jesli nawet bedzie to powinna byc specjalna klasa ktora odpowiada za zarzadzanie nimi( i tylko daje dostep do wizardow obiektom ktore potrzebuja tego). 
A i Wizard to bedzie Wizard a nie GameObject.
PS Wizard przez jedno z.

czkafa commented on Assets/Scripts/Miscellaneous/Explosion.cs:
```diff
@@ -0,0 +1,27 @@
+    private GameObject[] Wizzards;
+
+    public Explosion() {
+        Wizzards = GameObject.FindGameObjectsWithTag("Character");
```
Tak jak notka wyzej, nie potrzebujemy szukac Wizardow wsrod wszystkich obiektow(tylko nam pogorszy performance gry).
Mozna ewentualnie stworzyc singletona jakiegos ktory by przechowywal podstawowe informace o bitwie i z niego pobierac informacje na temat wizardow.

czkafa commented on Assets/Scripts/Miscellaneous/Explosion.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        foreach (GameObject wizard in Wizzards)
+        {
+            float distance = Vector3.Distance(wizard.transform.position, position);
```
W przypadku gdy bedzie Wizard za gorka ale w obszarze wybuchu tez powinien dostac obrazenia?

Elpicsida replied:
Dla uproszczenia zróbmy na razie analogicznie jak w Wormsach (obrażenia idą po okręgu bez względu na to czy jest teren czy nie).

czkafa commented on Assets/Scripts/Miscellaneous/Explosion.cs:
```diff
@@ -0,0 +1,27 @@
+
+    public void Explode(int damage, int radius, Vector3 position)
+    {
+        foreach (GameObject wizard in Wizzards)
```
Np gdyby byl ten menager wizardow mozna byloby tam dac delegatem akcje ktora mialaby byc zrobiona dla wszystkich Wizardow.

czkafa commented on Assets/Scripts/Miscellaneous/TurnManager.cs:
```diff
@@ -2,10 +2,10 @@
+public class TurnManager : MonoBehaviour {
 
-    public Transform[] Wizzards;
+    public GameObject[] Wizzards;
```
Tutaj powinny byc Wizardy a nie GameObjekty.
PS Wizard jedno z.

czkafa commented on Assets/Scripts/Miscellaneous/TurnManager.cs:
```diff
@@ -22,32 +22,38 @@ void Awake()
         {
-            Wizzards[0].gameObject.SendMessage("Activate");
-            Wizzards[1].gameObject.SendMessage("Deactivate");
+            Wizzards[0].SendMessage("Activate");
```
Powinienes uzywac jakiejs metody z Wizarda jesli chcesz w nim cos przestawic.

czkafa commented on Assets/Scripts/Spells/ShootingSpells/FireballBallistic.cs:
```diff
@@ -0,0 +1,25 @@
+
+    void OnCollisionEnter2D(Collision2D collision)
+    {
+        if (collision.gameObject.tag == "Terrain")
```
A gdy uderzy czarodzieja to nic sie nie stanie?

czkafa commented on Assets/Scripts/Wizzard/WizzardController.cs:
```diff
@@ -1,24 +1,25 @@
+    public float maxSpeed = 1f;
+    public int FireStrength = 500;
+    
+    private int currentSpell;
```
Jest to gdzies ustawiane?

Elpicsida replied:
Domyślnie int przyjmuje wartość 0 - także będzie ustawiony pierwszy czar z listy template :).

czkafa commented on Assets/Scripts/Wizzard/WizzardController.cs:
```diff
@@ -35,28 +36,23 @@ void Update()
+                    {
+                        Spell spell = fireball.GetComponent<Spell>();
+                        spell.Activate();
+                        wizardStats.Spells.Remove(wizardStats.Spells[currentSpell]);
```
Troche ta tablica speli jest dziwna. Spelle maja byc pokazywane na zasadzie A-spell1 B spell2 przykladowo ABBBBBAAAABBB zamiast A szt 5 B szt 8 Nie bedzie to zbyt czytelne dla uzytkownika chyba ze taki byl Twoj zamiar?

czkafa commented on Assets/Scripts/Wizzard/WizzardController.cs:
```diff
@@ -35,28 +36,23 @@ void Update()
+                }
+                else
+                {
+                    Debug.Log("The wizzard doesn't have any spells left");
```
Ten jak kilka innych Logów maja malo informacji, dodaj ktory to byl Wizard ktorego czaru mu brakowalo i itp.

czkafa commented on Assets/Scripts/Spells/Abstract/ShootingSpell.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public override void Activate()
+        {
+            TurnManager turnManager = GameObject.FindObjectOfType<TurnManager>();
```
Troche dziwne wydaje mi sie to by Szpel mial szukac samemu wlasciela i sie samemu wystrzeliwac.

Elpicsida replied:
Hmmm wydaje mi się, że tak jak jest ok :).
Spell jest wystrzeliwany z kontrolera , tam jest wywoływana metoda Active na zaklęciu.
Spelle mogą mieć różne zachowania (mogą korzystać z pozycji jednego z czarodziejów/ globalny/ wybór kursorem). Także jeśli miałbym to przenieść do klasy wizardController to musiałbym rozróżniać w zależności od typu czaru.

Elpicsida committed:
Small Refactoring
- Changed SendMessage to script method names
- Replaced GetComponent() by parameter reference

Elpicsida commented on Assets/Scripts/Spells/Abstract/ShootingSpell.cs:
```diff
@@ -8,19 +8,16 @@ namespace Assets.Scripts.Spells
-
         public override void Activate()
         {
             TurnManager turnManager = GameObject.FindObjectOfType<TurnManager>();
```
@czkafa
Starałem się to naprawić usuwając tę linijkę i wprowadzając parametr publiczny TurnManager. Przekazałem prefab TurnManager i wrzuciłem go na scenę ale leciał mi exception na klasach BattleGUI. Jak masz chęci mógłbyś to przerobić i wrzucić poprawki pod tym PRem ? :)

czkafa replied:
Wg mnie cala budowa tego szpela nie jest poprawana powinnien on byc wyrzucany na zewnatrz(przed aktywacja spela powinien on byc tworzony i dawana mu sila) czyli np dajesz wizardowi punkty z ktorego beda wystrzeliwane czary(zawsze zeby bylo z przodu) do spela albo do fabryki speli przychodza tylko argumenty sila i itp
